doc: add vCAT documentation (#6825)
projectacrn/acrn-hypervisor: doc/developer-guides/hld/hv-vcat.rst

.. _hv_vcat:

Enable vCAT
###########

vCAT refers to the virtualization of Cache Allocation Technology (CAT), one of the
RDT (Resource Director Technology) technologies.

ACRN vCAT is built on top of ACRN RDT: ACRN RDT provides a number of physical CAT resources
(CLOS IDs + cache ways), ACRN vCAT exposes some number of virtual CAT resources to VMs and
then transparently map them to the assigned physical CAT resources in the ACRN hypervisor;
VM can take advantage of vCAT to prioritize and partition virtual cache ways for its own tasks.

> In current CAT implementation, one CLOS ID corresponds to one IA32_type_MASK_n (type: L2 or L3,

Reviewer: ```suggestion
In current CAT implementation, one CLOS ID corresponds to one ``IA32_type_MASK_n`` (type: L2 or L3,
```

> n ranges from 0 to MAX_CACHE_CLOS_NUM_ENTRIES - 1) MSR and a bit in a capacity bitmask (CBM)

Reviewer: ```suggestion
n ranges from 0 to ``MAX_CACHE_CLOS_NUM_ENTRIES`` - 1) MSR and a bit in a capacity bitmask (CBM)
```

corresponds to one cache way.

> On current generation systems, normally L3 cache is shared by all CPUs on the same socket and

Reviewer: ```suggestion
On current generation systems, normally L3 cache is shared by all CPU cores on the same socket and
```
?

L2 cache is generally just shared by the hyperthreads on a core. But when dealing with ACRN
vCAT CLOS IDs assignment, it is currently assumed that all the L2/L3 caches (and therefore all CLOS IDs)
are system-wide caches shared by all cores in the system, this is done for convenience and to simplify
the vCAT configuration process. If vCAT is enabled for a VM (abbreviated as vCAT VM), there should not
be any CLOS ID overlap between a vCAT VM and any other VMs. e.g. the vCAT VM has exclusive use of the
assigned CLOS IDs.
When assigning cache ways, however, the VM can be given exclusive, shared, or mixed access to the cache
ways depending on particular performance needs. For example, use dedicated cache ways for RTVM, and use
shared cache ways between low priority VMs.

> In ACRN, vCAT can be configured via the "VM-Configuration". The vCAT cache ways allocated

Reviewer: Are you referring to the ACRN Configuration Tool?

Author: yes, referring to the vcat configuration

for VMs are determined in the VM configuration (:ref:`vcat_configuration`).

For further details on the RDT, refer to the ACRN RDT high-level design :ref:`hv_rdt`.

High Level ACRN vCAT Design
***************************

ACRN CAT virtualization support can be divided into two parts:

> - CAT Capability Exposure to Guest

Reviewer: When you say Guest, do you mean the User VMs **and** the Service VM? If so, then I would call it `Guest VM`. If the Service VM is not part of this, please use `User VM` instead.

Author: ok, will change to Guest VM

- CAT resources (CLOS IDs + cache ways) management

The figure below shows high-level design of vCAT in ACRN:

   .. figure:: images/vcat-hld.png
      :align: center

> CAT Capability Exposure to Guest

Reviewer: Same question regarding the use of `Guest`.

********************************

> ACRN exposes CAT capability and resource to a guest VM via vCPUID and vMSR, as explained

Reviewer: Ditto

in the following sections.

vCPUID
======

CPUID Leaf 07H
--------------

- CPUID.(EAX=07H, ECX=0).EBX.PQE[bit 15]: Supports RDT capability if 1. This bit will be set for a vCAT VM.

CPUID Leaf 10H
--------------

**CAT Resource Type and Capability Enumeration**

- CPUID.(EAX=10H, ECX=0):EBX[1]: If 1, indicate L3 CAT support for a vCAT VM.
- CPUID.(EAX=10H, ECX=0):EBX[2]: If 1, indicate L2 CAT support for a vCAT VM.
- CPUID.(EAX=10H, ECX=1): CAT capability enumeration sub-leaf for L3. Reports L3 COS_MAX and CBM_LEN to a vCAT VM
- CPUID.(EAX=10H, ECX=2): CAT capability enumeration sub-leaf for L2. Reports L2 COS_MAX and CBM_LEN to a vCAT VM

vMSR
====

The following CAT MSRs will be virtualized for a vCAT VM:

- IA32_PQR_ASSOC
- IA32_type_MASK_0 ~ IA32_type_MASK_n

> By default, after reset, all CPU cores are assigned to COS 0 and all IA32_type_MASK_n MSRs

Reviewer: Typo?
```suggestion
By default, after reset, all CPU cores are assigned to CLOS 0 and all IA32_type_MASK_n MSRs
```

Reviewer 2: Recommend using COS instead of CLOS. TCC team is going to start using COS in their next release, and SDM (Software Developer Manual) uses COS (mostly, there is some inconsistency), for example, "Bitmasks are configured via the IA32_resourceType_MASK_n MSRs, where resourceType indicates a resource type (e.g. “L3” for the L3 cache) and n indicates a COS number." https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3b-part-2-manual.pdf.

Author: OK, will change to COS

CAT resources (CLOS IDs + cache ways) management
************************************************

All accesses to the CAT MSRs are intercepted by vMSR and control is passed to vCAT, which will perform
the following actions:

- Intercept IA32_PQR_ASSOC MSR to re-map virtual CLOS ID to physical CLOS ID.
  Upon writes, store the re-mapped physical CLOS ID into its vCPU ``msr_store_area``
  data structure guest part. It will be loaded to physical IA32_PQR_ASSOC on each VM-Enter.

- Intercept IA32_type_MASK_n MSRs to re-map virtual CBM to physical CBM. Upon writes,
  program re-mapped physica CBM into corresponding physical IA32_type_MASK_n MSR

  Several vCAT P2V (physical to virtual) and V2P (virtual to physical)
  mappings exist, as illustrated in the following pseudocode:

.. code-block:: none

       struct acrn_vm_config *vm_config = get_vm_config(vm_id)

       max_pcbm = vm_config->max_type_pcbm (type: l2 or l3)
       mask_shift = ffs64(max_pcbm)

       vclosid = vmsr - MSR_IA32_type_MASK_0
       pclosid = vm_config->pclosids[vclosid]

       pmsr = MSR_IA32_type_MASK_0 + pclosid
       pcbm = vcbm << mask_shift
       vcbm = pcbm >> mask_shift

>        Where

Reviewer: We should take this part out of the `code-block`, and use double-backticks to highlight the the code-related items.

       vm_config->pclosids[]: array of physical CLOS IDs, where each corresponds to one vcpu_clos that
       is defined in the scenario file

       max_pcbm: a bitmask that selects all the physical cache ways assigned to the VM, corresponds to
       the nth CLOS_MASK that is defined in scenario file, where n = the first physical CLOS ID assigned
       = vm_config->pclosids[0]

       ffs64(value): find the first (least significant) bit set in value and return
       the index of that bit.

       MSR_IA32_type_MASK_0: 0xD10 for L2, 0xC90 for L3

       vclosid: virtual CLOS ID, always starts from 0

       pclosid: corresponding physical CLOS ID for a given vclosid

>        vmsr: virtual msr address, passed to vCAT handlers by the

Reviewer: ```suggestion
       vmsr: virtual MSR address, passed to vCAT handlers by the
```

>        caller functions rdmsr_vmexit_handler()/wrmsr_vmexit_handler()

Reviewer: ```suggestion
       caller functions ``rdmsr_vmexit_handler()``/``wrmsr_vmexit_handler()``
```
